util/log: protect against file name collisions during rotations.

knz opened the pull request:
Prior to this patch log files were named like this:

`cockroach.kenabook.kena.log.ERROR.2016-11-28T20_00_35Z.9524`

With this patch they are named like this:

`cockroach.kenabook.kena.2016-11-28T20_00_35Z.009524.ERROR.log`

Why?

The time stamp used to generate the file names is not guaranteed to be
monotonic (time can go backwards due to leap seconds or system
adjustments) or even to change fast enough that every new file gets a
different time stamp (time can be stalled by NTP to correct for a
skew).  The time going backwards is especially problematic, because it
can cause the apparent ordering of log files (based on their name) to
become incompatible with the actual ordering of the logging output.

How this was fixed / what's new?

- the numeric fields are formatted with a fixed width so that
  lexicographic sorting does the right thing.

- the file name extension `.log` is at the end so that utility
  programs run by users can guess what to do with a log file by just
  looking at its name.

Fixes #10936.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/11666)
<!-- Reviewable:end -->

knz committed:
util/log: change the naming of log files.

Prior to this patch log files were named like this:

`cockroach.kenabook.kena.log.ERROR.2016-11-28T20_00_35Z.9524`

With this patch they are named like this:

`cockroach.kenabook.kena.2016-11-28T20_00_35Z.009524.ERROR.log`

Why?

The time stamp used to generate the file names is not guaranteed to be
monotonic (time can go backwards due to leap seconds or system
adjustments) or even to change fast enough that every new file gets a
different time stamp (time can be stalled by NTP to correct for a
skew). The time going backwards is especially problematic, because it
can cause the apparent ordering of log files (based on their name) to
become incompatible with the actual ordering of the logging output.

How this was fixed / what's new?

- the numeric fields are formatted with a fixed width so that
  lexicographic sorting does the right thing.

- the file name extension `.log` is at the end so that utility
  programs run by users can guess what to do with a log file by just
  looking at its name.

knz commented:
I'm not happy about this PR. Now I'm thinking that if the clock goes back, the order of files will be messed up and the user may have no idea this happened.

What do you think about adding a file counter?

tamird commented:
Yeah, I think maybe that's the only reasonable thing to do - prefix each file with a monotonic counter.

cc @mberhault who may have opinions.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/util/log/file.go, line 220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KXgfhphzBjst5rPNHA6:-KXgfhphzBjst5rPNHA7:b40y1ve) ([raw file](https://github.com/cockroachdb/cockroach/blob/d926cb66a9ad26bc9657ff4e8ddfc297df659dcc/pkg/util/log/file.go#L220)):*
> ```Go
> 	fname := baseFname
> 	// If we are rotating into a new file within the same second, or
> 	// there is a NTP delay underway, or a leap second adjustment
> ```

fun fact: also daylight savings.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-:-KXgfqC_xEjNfJoX5Ifk:bey7o10)*
<!-- Sent from Reviewable.io -->

knz commented:
ok, done. PTAL.

tamird commented:
Well, this is a bit naive - you will still need to detect duplicates, or else restarting processes will not work?

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-:-KXghvL2hxaInNFYxJxp:b-98bssb)*
<!-- Sent from Reviewable.io -->

tamird commented:
Ah, you put the PID first. Hm, that breaks entirely the ability to sort these files, doesn't it?

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-:-KXgi4mIIR04twRjfS1L:b-m3rfl8)*
<!-- Sent from Reviewable.io -->

knz commented:
What do you mean sort? If you want to ignore the PID and still sort by time stamp you can do `sort -t. -k6`.

But this PR just established that sorting by time stamp is borked so you wouldn't want to do that anyway. Sorting by counter is really what you want but we don't persist this counter across restarts so you can't sort using that without the PID.

If we really want something more monotonic we need to persist the counter across node restarts. Do we want to go there?

knz commented:
Perhaps @bdarnell has an opinion here too.

knz commented:
Alternatively we could hook up the HLC to generate logging time stamps (and time stamps for log file names) and use the logical clock also in file names. Then the separate counter probably wouldn't be needed.

RaduBerinde commented:
I don't see any reason to put the PID and counter first. If anything it should come after the timestamp, so they still sort fine in normal circumstances. And by "sort" I mean the natural order when you list them, not `sort -t '.' -k6` (who's gonna type that every time they list log files?)

The current timestamp `2016-11-28T20_00_35Z.9524` is pretty standard and easier to read than `20161128_140035`

Is this a real problem in practice? I wouldn't change this just to fix a flaky test. The test could just wait in a loop until `time.Now()` is 1 second larger instead of sleeping 1 second.

RaduBerinde commented:
We could also remember the last rotation timestamp and use `MIN(time.Now(), last_timestamp + 1 second)` every time we rotate. This would guarantee ordering within a process (and we already add the pid too) and we wouldn't need to sleep at all in the test.

bdarnell commented:
I agree with @raduberinde; I think it's better and more expected to put the timestamp before the PID so that things sort chronologically in the normal case when clocks are well-behaved, and I'd prefer a test-only change to fix the flaky test. Outside of stress tests we shouldn't see log files turning over so quickly that clock behavior matters.

I also think that the time zone used for timestamps in filenames should match that used for the timestamps in the log lines themselves. If someone hasn't configured their server to use UTC then we should recognize that preference and log in their configured time zone. (even though DST switches are the most likely way to get out-of-order log files)

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-:-KXo49q9mQpse-Uxr9Qw:b9dy1e3)*
<!-- Sent from Reviewable.io -->

knz commented:
I think I have addressed the various counters:
- I have applied the "min" idea from Radu to ensure the timestamps are monotonic (inside the same process)
- I have reverted to the previous timestamp format, without a change in time zone (note that the current/previous implementation uses Go's default time zone, not what the user has configured. That happens to be UTC currently.)
- the test does not need to sleep.

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-:-KYE2TiWMzCmkn-osCis:b6kqus1)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/file.go, line 159 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KYEK6CEMiakDuee5mVp:-KYEK6CEMiakDuee5mVq:btl2pwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf70d31983ac67355ae25d9d9c8d443d8a7cd1b/pkg/util/log/file.go#L159)):*
> ```Go
> 	tFormatted := strings.Replace(t.Format(time.RFC3339), ":", "_", -1)
> 
> 	fCnt := atomic.AddInt64(&fileNameCounter, 1)
> ```

Why do we need this counter? We won't have duplicate filenames within the same process anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/log/file.go, line 159 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KYEK6CEMiakDuee5mVp:-KYEqXYJRM3IDuPPr97O:b-3wxtuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbf70d31983ac67355ae25d9d9c8d443d8a7cd1b/pkg/util/log/file.go#L159)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why do we need this counter? We won't have duplicate filenames within the same process anyway.
</blockquote></details>

Oh that's right, thanks to your earlier suggestion. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/util/log/file.go, line 162 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KYFPLbFGaX4QXVaDsia:-KYFPLbFGaX4QXVaDsib:b-21p5ps) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88d86a591b115c351c9479529f42ab972777a55/pkg/util/log/file.go#L162)):*
> ```Go
> 		tFormatted,
> 		pid,
> 		severity.Name())
> ```

Do we need to worry about any scripts somewhere using the old format? We could also just keep it as is since we're not adding anything

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-:-KYFPVqBa78kHRJ2oOJF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/util/log/file.go, line 162 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KYFPLbFGaX4QXVaDsia:-KYFRRJik-fozZopx5pr:b3sk8ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88d86a591b115c351c9479529f42ab972777a55/pkg/util/log/file.go#L162)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we need to worry about any scripts somewhere using the old format? We could also just keep it as is since we're not adding anything
</blockquote></details>

I like that the output of "ls" is sorted by timestamp first, instead of grouping the log levels together.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666)*
<!-- Sent from Reviewable.io -->

knz commented:
@tamird is this still good to go for you?

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 6 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/server/status_test.go, line 233 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KYFSXmq1v3N5_ePQhOC:-KYFSXmq1v3N5_ePQhOD:b9aanbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88d86a591b115c351c9479529f42ab972777a55/pkg/server/status_test.go#L233)):*
> ```Go
> 		for i, name := range []string{"ERROR.log", "INFO.log", "WARNING.log"} {
> 			if strings.Contains(wrapper.Files[i].Name, name) {
> 				found = true
> ```

optional: break

---

*[pkg/util/log/file.go, line 105 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KYFSo0V3xRk_VbEg0gl:-KYFSo0V3xRk_VbEg0gm:b-x6icle) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88d86a591b115c351c9479529f42ab972777a55/pkg/util/log/file.go#L105)):*
> ```Go
> // logFileRE matches log files to avoid exposing non-log files accidentally
> // and it splits the details of the filename into groups for easy parsing.
> // The log file format is {process}.{host}.{username}.{timestamp}.{pid}.{severity}.log
> ```

isn't this stale now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-:-KYFT64xnO8XuYfLeTfV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/server/status_test.go, line 233 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KYFSXmq1v3N5_ePQhOC:-KYFUHGXr6kKppcQRquu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88d86a591b115c351c9479529f42ab972777a55/pkg/server/status_test.go#L233)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

optional: break
</blockquote></details>

Done.

---

*[pkg/util/log/file.go, line 105 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666#-KYFSo0V3xRk_VbEg0gl:-KYFULuZdhCNdsZTxzzc:bv5q6if) ([raw file](https://github.com/cockroachdb/cockroach/blob/c88d86a591b115c351c9479529f42ab972777a55/pkg/util/log/file.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

isn't this stale now?
</blockquote></details>

No I have updated it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11666)*
<!-- Sent from Reviewable.io -->
